src/Sulu/Component/Webspace/StructureProvider/WebspaceStructureProvider.php

```diff
@@ -0,0 +1,137 @@
+ */
+class WebspaceStructureProvider implements WebspaceStructureProviderInterface
+{
+    /**
+     * @var \Twig_Environment
+     */
+    private $twig;
+
+    /**
+     * @var StructureManagerInterface
+     */
+    private $structureManager;
+
+    /**
+     * @var WebspaceManagerInterface
+     */
+    private $webspaceManager;
+
+    /**
+     * @var ActiveTheme
+     */
+    private $activeTheme;
+
+    /**
+     * @var Cache
+     */
+    private $cache;
+
+    /**
+     * WebspaceStructureProvider constructor.
+     *
+     * @param \Twig_Environment $twig
+     * @param StructureManagerInterface $structureManager
+     * @param WebspaceManagerInterface $webspaceManager
+     * @param ActiveTheme $activeTheme
+     * @param Cache $cache
+     */
+    public function __construct(
+        \Twig_Environment $twig,
+        StructureManagerInterface $structureManager,
+        WebspaceManagerInterface $webspaceManager,
+        ActiveTheme $activeTheme,
+        Cache $cache
+    ) {
+        $this->twig = $twig;
+        $this->structureManager = $structureManager;
+        $this->webspaceManager = $webspaceManager;
+        $this->activeTheme = $activeTheme;
+        $this->cache = $cache;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getStructures($webspaceKey)
+    {
+        if (!$this->cache->contains($webspaceKey)) {
+            return $this->loadStructures($webspaceKey);
+        }
+
+        $keys = $this->cache->fetch($webspaceKey);
+
+        return array_map(
+            function ($key) {
+                return $this->structureManager->getStructure($key);
+            },
+            $keys
+        );
+    }
+
+    private function loadStructures($webspaceKey)
+    {
+        $before = $this->activeTheme->getName();
+        $webspace = $this->webspaceManager->findWebspaceByKey($webspaceKey);
+        $this->activeTheme->setName($webspace->getTheme()->getKey());
+
+        $structures = array();
+        $keys = array();
+        foreach ($this->structureManager->getStructures() as $page) {
+            /** @var PageBridge $page */
+            $template = sprintf('%s.html.twig', $page->getView());
+            if ($this->templateExists($template)) {
+                $keys[] = $page->getKey();
+                $structures[] = $page;
+            }
+        }
+        $this->activeTheme->setName($before);
+
+        $this->cache->save($webspaceKey, $keys);
+
+        return $structures;
+    }
+
+    /**
+     * checks if a template with given name exists
+     *
+     * @param string $template
+     *
+     * @return bool
+     */
+    protected function templateExists($template)
+    {
+        $loader = $this->twig->getLoader();
+        if ($loader instanceof \Twig_ExistsLoaderInterface) {
+            return $loader->exists($template);
+        }
+
+        try {
+            $loader->getSource($template);
+
+            return true;
+        } catch (\Twig_Error_Loader $e) {
+        }
```

Comment:
Please add at least comment to the catch block. Or wouldn't you think it would be an even better idea to put the `return false` into the catch block? Then it would be clear that the `Twig_Error_Loader` exception causes the method to return false.